alpa/util.py

```diff
@@ -1243,3 +1255,247 @@ def try_import_ray_worker(error: bool = False):
+    placement group
+
+    Args:
+        num_hosts: the number of hosts to create the placement group for
+        num_devices_per_host: the number of devices per host
+        additional_resources_per_host: additional resources per host
+
+    Returns:
+        The placement group
+    """
+    current_placement_group = get_current_placement_group()
+    ray_worker = try_import_ray_worker()
+    worker = ray_worker.global_worker  # pylint: disable=protected-access
+    should_capture_child_tasks_in_placement_group = (
+        worker.should_capture_child_tasks_in_placement_group)
+    should_create_placement_group = (
+        current_placement_group is None or
+        not should_capture_child_tasks_in_placement_group)
+
+    if should_create_placement_group:
+        additional_resources_per_host = (additional_resources_per_host or {})
+        bundle = {
+            "CPU": 1,
+            "GPU": num_devices_per_host,
+            **additional_resources_per_host,
+        }
+        bundles = [bundle.copy() for _ in range(num_hosts)]
+
+        # Alpa Placement Group: `SPREAD` strategy is required
+        # https://docs.ray.io/en/latest/ray-core/placement-group.html#strategy-types
+        # Each bundle must be scheduled in a separate node.
+        strategy = "SPREAD"
+
+        placement_group = ray.util.placement_group(bundles, strategy=strategy)
+        logger.debug("Waiting for placement group to start.")
+        timeout = env_integer(PLACEMENT_GROUP_TIMEOUT_S_ENV, 100)
+        ready, _ = ray.wait([placement_group.ready()], timeout=timeout)
+        if ready:
+            logger.debug("Placement group has started.")
+        else:
+            raise TimeoutError(
+                "Placement group creation timed out. Make sure your "
+                "cluster either has enough resources or use an "
+                "autoscaling cluster. If you are running on a cluster, "
+                "make sure you specify an address in `ray.init()`, for example,"
+                ' `ray.init("auto")`. You can also increase the timeout by '
+                "setting the ALPA_PLACEMENT_GROUP_TIMEOUT_S environment "
+                "variable. Current resources available: "
+                f"{ray.available_resources()}, resources requested by "
+                f"the placement group: {placement_group.bundle_specs}")
+        return placement_group
+    else:
+        return current_placement_group
+
+
+def get_bundle_idx(placement_group, device_ips):
+    """Get the bundle index for the placement group.
+
+    The placement group is a list of resource bundles.
+    Each bundle will be assigned to a device.
+
+    First, we need to find the bundle index with GPU resources.
+    Then, we can find the device IP for the bundle index.
+    Lastly, we sort bundle index according to the device IP list given.
+
+    Args:
+        placement_group (placement group): The placement group.
+        device_ips (list): The list of device IP addresses.
```

Comment:
coding style. We add type annotations to the function signature instead of the docstring.